ENG-4356: Guide content for IdP attribute mapping and application complete registration (#4266)
FusionAuth/fusionauth-site: astro/src/content/docs/lifecycle/authenticate-users/identity-providers/index.mdx

* Use a reconcile lambda to ensure that the email address or identifier provided by the Identity Provider is expected.
* Use a `Disabled` linking strategy for an Identity Provider and manage links via the [Links API](/docs/apis/identity-providers/links); this allows business logic to execute before or during linking. Examples include disallowing any links to a certain domain or calling into another API for additional validation before creating a link.

> ## Attribute Mappings

Author: Should this add the same [aside](https://github.com/FusionAuth/fusionauth-site/pull/4254/changes/0c420a1a6494ded90e66689eea15102abaecfea5..2272cb6a53146746a18f7c0b5a7951e151f7f54e#diff-2f4bd3dc48ef0078d1da0b85b2cd32de3989f22b94bfa4aeb5be0f6434c50d91R82) as the application configuration section?

```
<Aside type="note">
Saving an identity provider in Tenant Manager reapplies Tenant Manager-managed values. This replaces `attributeMappings` and `applicationConfiguration` with Tenant Manager-managed values by design. See [Tenant Manager Identity Provider Restrictions](/docs/lifecycle/manage-users/tenant-manager#tenant-manager-identity-provider-restrictions).
</Aside>
```

Reviewer: Might as well!

<Aside type="version">
Available since 1.65.0.
</Aside>

> Attribute mappings allow you to map claims from an Identity Provider response directly to FusionAuth user and registration fields. They are a simpler no-code alternative to reconciliation lambdas. Attribute mappings are key value pairs that can be configured via the Identity Provider API or the Tenant Manager.

Author: ❗ I think we should link to APIs and TM SSO pages here once it comes together in the release branch.

Reviewer: What do you think of replacing "simpler no-code" with "declarative"? While "simpler no-code" isn't wrong, I would prefer to avoid simpler (it's a subjective term) and no-code is a bit jargon-y and might not age well.

Author: Much better. I felt icky saying "no-code"

Attribute mapping keys represent fields on a FusionAuth user or registration and must begin with `user.` or `registration.`. Some example keys are:
* `user.email` - The user's email address. This could be used for linking the user according to the configuring linking strategy.
* `registration.data.subscription` - Custom registration data with information on the user's subscription.

> Attribute mapping values represent fields or claims from the Identity Provider response. The shape of the response matches the parameter available in the corresponding reconcile lambda type. **Note:** The OpenID Connect reconcile lambda has two parameters, `jwt` and `id_token`, attribute mapping values must include one of these as a prefix in order to indicate which parameter they are referencing.

Reviewer: Instead of bold text, could we put this in an aside? It might qualify as a relatively important warning since that sounds like a OIDC-specific footgun?

> The values can be specified using a simple dot notation or JSON pointer notation. JSON pointer notation is necessary to refer to fields containing `.` or `/` characters. A literal `/` in a JSON pointer is escaped as `~1`. Some example values are:

Reviewer: ```suggestion
The values can be specified using dot or JSON pointer notation. To refer to a field that contains `.` or `/`, use JSON pointer notation. A literal `/` in a JSON pointer is escaped as `~1`. Some example values are:
```
Suggestion: slight rephrases for readability with slightly more prescriptive wording.

I'm not sure if there are any other special characters that require escapes here, but if they exist, we might want to break those escape sequences out into a bulleted list. But if it's just the one, this seems fine.

Author: The only other escape character is `~0` for a literal `~`. I'll add that for completeness.

* `id_token.preferred_username` - the `preferred_username` claim on the OIDC ID token using dot notation
* `/jwt/https:~1~1example.com~1favorite_color` - a custom claim `https://example.com/favorite_color` on the OIDC access token using JSON pointer notation
* `assertion.attributes.email.0` - The first email attribute value on a SAML response using dot notation. All SAML attributes are arrays, so accessing a single value requires providing an index. **Note:** the attribute mapping does not include the `samlResponse` lambda parameter name in its value. Only OIDC requires specifying the parameter name.
* `/assertion/attributes/email/0` - the first email attribute value on a SAML response using JSON pointer notation

During reconciliation, attribute mapping values are used to extract a claim value from the Identity Provider response. The corresponding key is used to set that value on the FusionAuth user or registration. Some other notes on the reconciliation process:
* Failures locating a value on the Identity Provider response are ignored
* Failures setting a value on the user or registration object are ignored

> * All attribute mappings are considered optional

Reviewer: ```suggestion
During reconciliation, FusionAuth reads claim values using the attribute mapping values declared in the Identity Provider response. FusionAuth assigns the retrieved values to the corresponding key on the FusionAuth user or registration. All attribute mappings are optional; if FusionAuth cannot retrieve a claim value or set a key, it omits the mapping, but still continues reconciliation.
```

Suggestion: slight reword for more active voice.
Simplification: I think my last sentence here sums up the important part of the removed bullet points? Personally I can't think of a situation where we'd fail to set a value (maybe a duplicate or an invalid name?), but I think the important part is the optionality (and the logging), so I've reworded it to emphasize those bits.

Author: I reworked the last sentence a bit for accuracy, but I'm open to suggestions
> All attribute mappings are optional; if FusionAuth cannot retrieve a claim value or set a key, it continues the reconciliation process without.

- We don't currently log attribute mapping failures
- It will continue the reconciliation process, but it may fail if we were not able to reconcile the linking identity type
- User registration from an IdP login is a separate configuration
- A failed mapping is likely due to either user error or the value not being present on the IdP response for some users

Reviewer: Updated my suggestion since we don't log, hopefully that works.

> Reconcile lambdas and attribute mappings are mutually exclusive. Attempting to configure both on an Identity Provider is a validation error.

Reviewer: ```suggestion
Reconcile lambdas and attribute mappings are mutually exclusive. You cannot configure both on a single Identity Provider.
```
Unless you think 'validation error' is a particularly useful term, I think it's better to make this sentence simpler. Up to you!
